### PR TITLE
[C++ API] Use ATen dropout implementation in Dropout module and add FeatureDropout

### PR DESCRIPTION
**Pull request by goldsborough**
This PR does two things:
1. Replaces the implementation of the `Dropout` module with a call to the ATen function,
2. Replaces `Dropout2d` with a new `FeatureDropout` module that shall take the place of `Dropout2d` and `Dropout3d`. I contemplated calling it `Dropout2d` and making `Dropout3d` an alias for it, but similar to our decision for `BatchNorm{1,2,3}d` (c.f. https://github.com/pytorch/pytorch/pull/9188), we can deviate from Python PyTorch in favor of the ideal-world solution, which is to have a single module, since both actually just call `feature_dropout`.

I also replaced the implementation of `dropout3d`  with a call to `dropout2d` in Python. The code is the same and it's easier for developers to parse than having to manually match the tokens to make sure it's really 100% the same code (which it is, if I matched the tokens correctly).

@ebetica @ezyang @SsnL

### CODE REVIEW

File: test/cpp/api/integration.cpp [hunk 1]
```diff
@@ -336,7 +336,7 @@ TEST_CASE("integration/mnist", "[cuda]") {
   auto conv1 = model->add(Conv2d(1, 10, 5), "conv1");
   auto conv2 = model->add(Conv2d(10, 20, 5), "conv2");
   auto drop = Dropout(0.3);
-  auto drop2d = Dropout2d(0.3);
+  auto drop2d = Dropout(0.3);
```
Comment (SsnL) [inline test/cpp/api/integration.cpp]: What happened here? Dropout isn't the same with Dropout2d.

Comment (goldsborough) [reply test/cpp/api/integration.cpp]: Oh, I somehow assumed we were figuring out whether to apply 1-d or 2-d dropout inside the `at::dropout` call because I didn't see a `at::dropout2d` Function. Looking at `functional.py` I now see this function is called `feature_dropout`. I'll fix this PR then and also add Dropout3d while I'm at it.

**Comment by goldsborough** [comment]
@SsnL are you ok with the Python change?

File: torch/nn/functional.py [hunk 2]
```diff
@@ -662,11 +662,7 @@ def dropout3d(input, p=0.5, training=True, inplace=False):
         training: apply dropout if is ``True``. Defualt: ``True``
         inplace: If set to ``True``, will do this operation in-place. Default: ``False``
     """
-    if p < 0 or p > 1:
-        raise ValueError("dropout probability has to be between 0 and 1, "
-                         "but got {}".format(p))
-    f = _VF.feature_dropout_ if inplace else _VF.feature_dropout
-    return f(input, p, training)
+    return dropout2d(input, p, training, inplace)
```
Comment (SsnL) [inline torch/nn/functional.py]: I would not change to calling dropout2d here (yes I know it's the same code). It will be confusing if it errors and users see dropout2d in their trace.

**Review by ssnl — APPROVED** [review]
LOL nice!

File: test/cpp/api/integration.cpp (hunk 1 above)
Comment (SsnL) [reply test/cpp/api/integration.cpp]: Maybe not in this PR, but we should make feature dropout 1d to just call dropout in aten.

**Review by ebetica — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
comment looks fine
